crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -38,77 +41,268 @@ impl Pruner {
+                                .update_watermarks(vec![Watermark::new_lower_bound_tail(
+                                    WatermarkEntity::Checkpoints,
+                                    watermark.pruner_lo(),
+                                )])
+                                .await?;
+                        }
+                        WatermarkEntity::EvLookup => {
+                            self.store
+                                .prune_event_indices_table(
+                                    watermark.pruner_lo(),
+                                    inclusive_upper_bound,
+                                )
+                                .await?;
+                            self.store
+                                .update_watermarks(vec![Watermark::new_lower_bound_tail(
+                                    WatermarkEntity::EvLookup,
+                                    inclusive_upper_bound,
+                                )])
+                                .await?;
+                        }
+                        WatermarkEntity::TxLookup => {
+                            self.store
+                                .prune_tx_indices_table(
+                                    watermark.pruner_lo(),
+                                    inclusive_upper_bound,
+                                )
+                                .await?;
+                            self.store
+                                .update_watermarks(vec![Watermark::new_lower_bound_tail(
+                                    WatermarkEntity::TxLookup,
+                                    inclusive_upper_bound,
+                                )])
+                                .await?;
+                        }
+                        _ => {}
                     }
-                    self.partition_manager
-                        .drop_table_partition(table_name.clone(), epoch)
-                        .await?;
-                    info!(
-                        "Batch dropped table partition {} epoch {}",
-                        table_name, epoch
-                    );
                 }
             }
+        }
+        info!("Pruner task cancelled.");
+        Ok(())
+    }
+}
+
+/// Check if the lowest epoch of unpruned data exceeds the retention policy.
+fn should_update_watermark(watermark: &WatermarkRead, epochs_to_keep: u64) -> bool {
+    watermark.epoch_lo + epochs_to_keep <= watermark.epoch_hi
+}
+
+/// Determine if pruning should occur based on the current watermark state. When an entity's `hi`
+/// and `lo` (`reader_lo`) watermarks are updated to reflect the latest data range, if the
+/// `reader_lo` value is at least 1 greater than `pruned_lo`, then the pruner should prune. When
+/// `lo` is updated, and `pruned_lo` is still unset, that also means the pruner should start
+/// pruning.
+fn should_prune(watermark: &WatermarkRead) -> bool {
+    match watermark.pruned_lo {
+        None => watermark.reader_lo() > 0,
```

Comment:
When can this be `None`? And why do we prune if this is `None`?

Reply:
When the writer starts committing, it doesn't populate the `pruned_lo` field, but sets `lo` to 0. Similarly, committer sets `epoch_hi`, and `epoch_lo` defaults to 0. Once an entry has been inserted, committer updates only the `epoch_hi` and `hi` columns.

The pruner knows to update the watermark (both the `epoch_` and its range `lo` `hi`) when `epoch_hi - (epochs-to-keep - 1) > epoch_lo`

The pruner knows to update lookup tables when `lo > 0` but `pruned_lo` hasn't been set it, because `lo > 0` is true when pruner updates the watermark. If `pruned_lo` is None, that means there's pruning work to be done but pruner hasn't started yet

And once `pruned_lo` is set, as long as it is `pruned_lo < lo + 1`, we know that there is more data to prune. `pruned_lo` represents the latest pruned data, and `lo` is the smallest available data, so we are caught up when `pruned_lo + 1 = lo`